File: badge_generator.js
// badge_generator.js
const { program } = require('commander');
const QRCode = require('qrcode');
const sharp = require('sharp');
const fs = require('fs');
const csv = require('csv-parser');

class BadgeGenerator {
    constructor(options) {
        this.data = options.data;
        this.name = options.name || options.data;
        this.title = options.title || '';
        this.company = options.company || '';
        this.size = options.size || 200;
        this.color = options.color || '#000000';
        this.logoPath = options.logo;
        this.template = options.template ? JSON.parse(fs.readFileSync(options.template)) : this.defaultTemplate();
    }

    defaultTemplate() {
        return {
            badgeWidth: 600,
            badgeHeight: 400,
            background: '#FFFFFF',
            nameFontSize: 40,
            titleFontSize: 24,
            companyFontSize: 28,
            qrPosition: [380, 80],
            textPosition: [50, 100],
            textColor: '#000000'
        };
    }

    async createQR() {
        // Generate QR as buffer
        const qrBuffer = await QRCode.toBuffer(this.data, {
            errorCorrectionLevel: 'H',
            width: this.size,
            color: {
                dark: this.color,
                light: '#FFFFFF'
            },
            margin: 2
        });
        let img = sharp(qrBuffer);
        // Embed logo if provided
        if (this.logoPath) {
            const logo = sharp(this.logoPath);
            const logoSize = Math.round(this.size * 0.25);
            const logoBuffer = await logo.resize(logoSize, logoSize).toBuffer();
            // Overlay logo on QR
            const overlay = await sharp({
                create: {
                    width: this.size,
                    height: this.size,
                    channels: 4,
                    background: { r: 0, g: 0, b: 0, alpha: 0 }
                }
            }).composite([{ input: logoBuffer, gravity: 'centre' }]).png().toBuffer();
            img = sharp(qrBuffer).composite([{ input: overlay }]);
        }
        return img.toBuffer();
    }

    async generate(outputPath) {
        const template = this.template;
        // Create badge background
        const badgeWidth = template.badgeWidth;
        const badgeHeight = template.badgeHeight;
        // We'll use sharp to build the badge
        let badge = sharp({
            create: {
                width: badgeWidth,
                height: badgeHeight,
                channels: 3,
                background: template.background
            }
        });

        // Create text layers using SVG
        let svgText = `<svg width="${badgeWidth}" height="${badgeHeight}">`;
        const textColor = template.textColor;
        const [textX, textY] = template.textPosition;
        svgText += `<text x="${textX}" y="${textY}" font-size="${template.nameFontSize}" fill="${textColor}" font-family="Arial">${this.name}</text>`;
        if (this.title) {
            svgText += `<text x="${textX}" y="${textY + template.nameFontSize + 10}" font-size="${template.titleFontSize}" fill="${textColor}" font-family="Arial">${this.title}</text>`;
        }
        if (this.company) {
            const yOff = template.nameFontSize + template.titleFontSize + 20;
            svgText += `<text x="${textX}" y="${textY + yOff}" font-size="${template.companyFontSize}" fill="${textColor}" font-family="Arial">${this.company}</text>`;
        }
        svgText += `</svg>`;

        // Generate QR buffer
        const qrBuffer = await this.createQR();

        // Combine: overlay QR and text onto badge
        const [qrX, qrY] = template.qrPosition;
        const qrImage = await sharp(qrBuffer).resize(this.size, this.size).toBuffer();

        badge = badge.composite([
            { input: Buffer.from(svgText), top: 0, left: 0 },
            { input: qrImage, top: qrY, left: qrX }
        ]);

        // Save
        await badge.toFile(outputPath);
        console.log(`Badge saved to ${outputPath}`);
    }
}

program
    .requiredOption('-d, --data <text>', 'Data for QR code')
    .option('-n, --name <text>', 'Name on badge')
    .option('-t, --title <text>', 'Title')
    .option('-c, --company <text>', 'Company')
    .option('-o, --output <file>', 'Output file', 'badge.png')
    .option('-s, --size <number>', 'QR size', parseInt, 200)
    .option('--color <color>', 'QR color', '#000000')
    .option('--logo <path>', 'Logo path')
    .option('--template <path>', 'Template JSON file')
    .option('--batch <csv>', 'CSV for batch generation')
    .option('--output-dir <dir>', 'Output directory for batch', '.')
    .parse(process.argv);

const opts = program.opts();

if (opts.batch) {
    // Batch mode
    const results = [];
    fs.createReadStream(opts.batch)
        .pipe(csv())
        .on('data', (row) => results.push(row))
        .on('end', async () => {
            for (const row of results) {
                const data = row.data || '';
                const name = row.name || data;
                const title = row.title || '';
                const company = row.company || '';
                const outName = row.output || `${name.replace(/\s/g, '_')}.png`;
                const outPath = `${opts.outputDir}/${outName}`;
                const gen = new BadgeGenerator({
                    data, name, title, company,
                    size: opts.size,
                    color: opts.color,
                    logo: opts.logo,
                    template: opts.template
                });
                await gen.generate(outPath);
            }
        });
} else {
    const gen = new BadgeGenerator({
        data: opts.data,
        name: opts.name || opts.data,
        title: opts.title,
        company: opts.company,
        size: opts.size,
        color: opts.color,
        logo: opts.logo,
        template: opts.template
    });
    gen.generate(opts.output).catch(console.error);
}

File: badge_generator.py

---

# Код на 8 языках программирования

## 1. Python (`badge_generator.py`)

```python
# badge_generator.py
import argparse
import csv
import json
import os
from PIL import Image, ImageDraw, ImageFont
import qrcode
from qrcode.image.styledpil import StyledPilImage
from qrcode.image.styles.moduledrawers import RoundedModuleDrawer
from qrcode.image.styles.colormasks import SolidFillColorMask
import io

class BadgeGenerator:
    def __init__(self, data, name, title=None, company=None, size=200,
                 color="#000000", logo_path=None, template=None):
        self.data = data
        self.name = name if name else data
        self.title = title
        self.company = company
        self.size = size
        self.color = color
        self.logo_path = logo_path
        self.template = self.load_template(template) if template else self.default_template()

    def default_template(self):
        return {
            "badge_width": 600,
            "badge_height": 400,
            "background": "#FFFFFF",
            "font_path": None,  # uses default
            "name_font_size": 40,
            "title_font_size": 24,
            "company_font_size": 28,
            "qr_position": (380, 80),
            "text_position": (50, 100),
            "text_color": "#000000"
        }

    def load_template(self, path):
        with open(path, 'r') as f:
            return json.load(f)

    def create_qr(self):
        qr = qrcode.QRCode(
            version=1,
            error_correction=qrcode.constants.ERROR_CORRECT_H,
            box_size=10,
            border=4,
        )
        qr.add_data(self.data)
        qr.make(fit=True)
        # Create styled image
        img = qr.make_image(
            image_factory=StyledPilImage,
            module_drawer=RoundedModuleDrawer(),
            color_mask=SolidFillColorMask(
                back_color=(255, 255, 255),
                front_color=self.color
            )
        ).convert('RGB')
        img = img.resize((self.size, self.size), Image.Resampling.LANCZOS)
        # Embed logo if provided
        if self.logo_path:
            logo = Image.open(self.logo_path).convert("RGBA")
            logo_size = int(self.size * 0.25)
            logo = logo.resize((logo_size, logo_size), Image.Resampling.LANCZOS)
            pos = ((self.size - logo_size) // 2, (self.size - logo_size) // 2)
            img.paste(logo, pos, logo)
        return img

    def generate(self, output_path):
        template = self.template
        # Create base badge
        badge = Image.new('RGB', (template['badge_width'], template['badge_height']),
                          template['background'])
        draw = ImageDraw.Draw(badge)

        # Load font (default if None)
        try:
            font_name = ImageFont.truetype(template.get('font_path', 'arial.ttf'),
                                           template['name_font_size'])
            font_title = ImageFont.truetype(template.get('font_path', 'arial.ttf'),
                                            template['title_font_size'])
            font_company = ImageFont.truetype(template.get('font_path', 'arial.ttf'),
                                              template['company_font_size'])
        except:
            font_name = ImageFont.load_default()
            font_title = font_name
            font_company = font_name

        # Draw text
        text_x, text_y = template['text_position']
        draw.text((text_x, text_y), self.name, fill=template['text_color'], font=font_name)
        if self.title:
            draw.text((text_x, text_y + template['name_font_size'] + 10),
                      self.title, fill=template['text_color'], font=font_title)
        if self.company:
            draw.text((text_x, text_y + template['name_font_size'] + template['title_font_size'] + 20),
                      self.company, fill=template['text_color'], font=font_company)

        # Paste QR
        qr_img = self.create_qr()
        qr_x, qr_y = template['qr_position']
        badge.paste(qr_img, (qr_x, qr_y))

        # Save
        badge.save(output_path)
        print(f"Badge saved to {output_path}")

def batch_generate(csv_path, output_dir, **kwargs):
    with open(csv_path, 'r') as f:
        reader = csv.DictReader(f)
        for row in reader:
            data = row.get('data', '')
            name = row.get('name', data)
            title = row.get('title', '')
            company = row.get('company', '')
            out_name = row.get('output', f"{name.replace(' ', '_')}.png")
            out_path = os.path.join(output_dir, out_name)
            gen = BadgeGenerator(data, name, title, company, **kwargs)
            gen.generate(out_path)

def main():
    parser = argparse.ArgumentParser(description="Генератор бейджей с QR-кодом")
    parser.add_argument("--data", required=True, help="Данные для QR-кода")
    parser.add_argument("--name", help="Имя на бейдже (если не указано, используется data)")
    parser.add_argument("--title", help="Должность")
    parser.add_argument("--company", help="Компания")
    parser.add_argument("--output", default="badge.png", help="Выходной файл (PNG/PDF)")
    parser.add_argument("--size", type=int, default=200, help="Размер QR-кода (пиксели)")
    parser.add_argument("--logo", help="Путь к логотипу для вставки в QR")
    parser.add_argument("--color", default="#000000", help="Цвет QR-кода (HEX)")
    parser.add_argument("--template", help="Файл шаблона (JSON)")
    parser.add_argument("--batch", help="CSV-файл для пакетной генерации")
    parser.add_argument("--output-dir", default=".", help="Директория для пакетного вывода")
    args = parser.parse_args()

    if args.batch:
        # Пакетный режим
        kwargs = {
            'size': args.size,
            'color': args.color,
            'logo_path': args.logo,
            'template': args.template
        }
        batch_generate(args.batch, args.output_dir, **kwargs)
    else:
        gen = BadgeGenerator(
            data=args.data,
            name=args.name or args.data,
            title=args.title,
            company=args.company,
            size=args.size,
            color=args.color,
            logo_path=args.logo,
            template=args.template
        )
        gen.generate(args.output)

if __name__ == "__main__":
    main()
